### PR TITLE
Recommend Prophecy for test doubles creation

### LINKED ISSUE #89: Recommend using Prophecy for mocking
Prophecy comes with phpunit [since version 4.5](https://thephp.cc/news/2015/02/phpunit-4-5-and-prophecy) , and I think we should default to using it for mocking things, because it is way way more concise unless you are mocking a class with a no arguments constructor and adding no expectations.

So several questions follow this : 
1. Should we allow using prophecy in tests : I vote :+1: 
2. Should we recommend using prophecy in new tests : I vote :+1: 
3. Should we refuse phpunit mocks in new tests : I vote :-1: 
4. Should we migrate all tests to prophecy, for consistency's sake  I vote :-1:
5. Should we refuse mixing Prophecy and Mocks in the same test (method or class) :+1:

### PR DESCRIPTION
**Pull request by greg0ire**
Fixes #89.

### CODE REVIEW

File: project/CONTRIBUTING.md [hunk 1]
```diff
@@ -97,6 +97,19 @@ Some rules have to be respected about the test:
 * Most of the time, the test class should have the same name as the targeted class, suffixed by `Test`.
 * The `@expectedException*` annotations are prohibited. Use `PHPUnit_Framework_TestCase::setExpectedException()`.
 
+##### Using test doubles
+
+Since phpunit 4.5, phpunit requires and [integrates](https://phpunit.de/manual/current/en/test-doubles.html#test-doubles.prophecy)
+the [phpspec/prophecy](https://github.com/phpspec/prophecy).
+Historically, Sonata has been using [the built-in test doubles implementation](https://phpunit.de/manual/current/en/test-doubles.html),
+but has deciding to move to Prophecy, which is more concise than its built-in counterpart is most cases.
```
Comment (OskarStark) [inline project/CONTRIBUTING.md]: maybe link to the discussion, if someone is interested?

File: project/CONTRIBUTING.md [hunk 2]
```diff
@@ -97,6 +97,19 @@ Some rules have to be respected about the test:
...
+This means the current Sonata codebase currently uses both implementations.
+If you want to contribute a test that uses test doubles, please follow these rules :
+
+1. All new test classes MUST use Prophecy.
+2. If you are changing an existing test class, you MUST use the same implementation it already uses.
```
Comment (OskarStark) [inline project/CONTRIBUTING.md]: ...., to keep the diff small.     ❓

File: project/CONTRIBUTING.md [hunk 3]
```diff
@@ -97,6 +97,19 @@ Some rules have to be respected about the test:
...
+3. You MAY submit a PR that migrates a test class from the built-in test double implementation to Prophecy.
```
Comment (OskarStark) [inline project/CONTRIBUTING.md]: ... a separate PR ...

**Comment by OskarStark** [comment]
LGTM 👍

File: project/CONTRIBUTING.md (hunk 3 above)
Comment (Soullivaneuh) [inline project/CONTRIBUTING.md]: a separate**d** PR. :stuck_out_tongue:

File: project/CONTRIBUTING.md (hunk 2 above)
Comment (Soullivaneuh) [inline project/CONTRIBUTING.md]: Or simply: To focus on the PR goal only.

File: project/CONTRIBUTING.md (hunk 3 above)
Comment (greg0ire) [inline project/CONTRIBUTING.md]: > a separated PR. :stuck_out_tongue:

Nope

File: project/CONTRIBUTING.md (hunk 2 above)
Comment (greg0ire) [inline project/CONTRIBUTING.md]: Oh. What about if this is a new method? I'll add a separate (no d) bullet point.

File: project/CONTRIBUTING.md (hunk 3 above)
Comment (Soullivaneuh) [inline project/CONTRIBUTING.md]: Why no? It should be on a dedicated PR. If this about add a feature, the class should just not be changed.

File: project/CONTRIBUTING.md (hunk 2 above)
Comment (Soullivaneuh) [inline project/CONTRIBUTING.md]: > What about if this is a new method?

Didn't understand, could you please elaborate?

File: project/CONTRIBUTING.md (hunk 3 above)
Comment (greg0ire) [inline project/CONTRIBUTING.md]: I'll add a sentence to clarify what you mean instead of adding this word, because people might start to ask "separate from one".

**Comment by greg0ire** [comment]
@OskarStark : I changed some things, please review it again entirely.

File: project/CONTRIBUTING.md [hunk 4]
```diff
@@ -97,6 +97,24 @@ Some rules have to be respected about the test:
 * Most of the time, the test class should have the same name as the targeted class, suffixed by `Test`.
 * The `@expectedException*` annotations are prohibited. Use `PHPUnit_Framework_TestCase::setExpectedException()`.
 
+##### Using test doubles
+
+Since phpunit 4.5, phpunit requires and [integrates](https://phpunit.de/manual/current/en/test-doubles.html#test-doubles.prophecy)
```
Comment (Soullivaneuh) [inline project/CONTRIBUTING.md]: phpunit -> PHPUnit

Because, it is it name. :stuck_out_tongue:

File: project/CONTRIBUTING.md [hunk 5]
```diff
@@ -97,6 +97,24 @@ Some rules have to be respected about the test:
...
+but [has decided to move to Prophecy](https://github.com/sonata-project/dev-kit/issues/89),
+which is more concise than its built-in counterpart is most cases.
+This means the current Sonata codebase currently uses both implementations.
+If you want to contribute a test that uses test doubles, please follow these rules :
+
+1. All new test classes MUST use Prophecy.
+2. If you are changing an existing test method, you MUST use the same implementation it already uses,
+and focus on the goal of your PR and only on that.
+3. If you are changing an existing test class, you MUST use the same implementation it already uses,
+especially if there is a `setUp()` method.
```
Comment (Soullivaneuh) [inline project/CONTRIBUTING.md]: Why especially? If you have some mock on a class, just use mock, not both.

Comment (greg0ire) [inline project/CONTRIBUTING.md]: What reason should I put then?

Comment (greg0ire) [inline project/CONTRIBUTING.md]: Replaced that with "to be more consistent", for lack of a better reason.

File: project/CONTRIBUTING.md (hunk 3 above)
Comment (greg0ire) [inline project/CONTRIBUTING.md]: http://english.stackexchange.com/questions/13144/separated-versus-separate

**Comment by OskarStark** [comment]
Good job @greg0ire, i like it 👍

File: project/CONTRIBUTING.md [hunk 6]
```diff
@@ -97,6 +97,24 @@ Some rules have to be respected about the test:
 * Most of the time, the test class should have the same name as the targeted class, suffixed by `Test`.
 * The `@expectedException*` annotations are prohibited. Use `PHPUnit_Framework_TestCase::setExpectedException()`.
 
+##### Using test doubles
+
+Since version 4.5, PHPUnit requires and [integrates](https://phpunit.de/manual/current/en/test-doubles.html#test-doubles.prophecy)
```
Comment (OskarStark) [inline project/CONTRIBUTING.md]: should we link to "current" or version 4.5 explicitely?

File: project/CONTRIBUTING.md [hunk 7]
```diff
@@ -97,6 +97,24 @@ Some rules have to be respected about the test:
...
+the [phpspec/prophecy](https://github.com/phpspec/prophecy).
+Historically, Sonata has been using [the built-in test doubles implementation](https://phpunit.de/manual/current/en/test-doubles.html),
```
Comment (OskarStark) [inline project/CONTRIBUTING.md]: here the "current" is ok IMO

File: project/CONTRIBUTING.md (hunk 6 above)
Comment (greg0ire) [inline project/CONTRIBUTING.md]: We should link to current b/c we are using the current version in our software. No need to link to outdated instructions

File: project/CONTRIBUTING.md (hunk 2 above)
Comment (greg0ire) [inline project/CONTRIBUTING.md]: If we have a test class full of mocks, and we want to add a new test method, what should the rules be? Is it ok to use prophecy or not? If not, why?

**Comment by jordisala1991** [comment]
LGTM, we should start moving to prophecy 👍

**Comment by greg0ire** [comment]
@Soullivaneuh this is 6 months old. What's the deal?
